hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/common/CalcIntoScanPushdownRule.java

```diff
@@ -33,61 +36,66 @@
  * See {@link HazelcastTable} for more information about constrained scans.
  * <p>
  * Before:
  * <pre>
- * LogicalCalc[filter=exp1]
- *     LogicalScan[table[filter=exp2]]
+ * Calc[filter=exp1]
+ *     TableScan[table[filter=exp2]]
  * </pre>
  * After:
  * <pre>
- * LogicalScan[table[filter=exp1 AND exp2]]
+ * TableScan[table[filter=exp1 AND exp2]]
  * </pre>
  */
 @Value.Enclosing
-public final class CalcIntoScanLogicalRule extends RelRule<Config> implements TransformationRule {
+public final class CalcIntoScanPushdownRule extends RelRule<Config> implements TransformationRule {
 
     @Value.Immutable
     public interface Config extends RelRule.Config {
-        CalcIntoScanLogicalRule.Config DEFAULT = ImmutableCalcIntoScanLogicalRule.Config.builder()
-                .description(CalcIntoScanLogicalRule.class.getSimpleName())
+        CalcIntoScanPushdownRule.Config DEFAULT = ImmutableCalcIntoScanPushdownRule.Config.builder()
+                .description(CalcIntoScanPushdownRule.class.getSimpleName())
                 .operandSupplier(b0 -> b0
-                        .operand(CalcLogicalRel.class)
+                        .operand(Calc.class)
                         .trait(LOGICAL)
                         .inputs(b1 -> b1
-                                .operand(FullScanLogicalRel.class).anyInputs()))
+                                .operand(FullScan.class).anyInputs()))
                 .build();
 
         @Override
         default RelOptRule toRule() {
-            return new CalcIntoScanLogicalRule(this);
+            return new CalcIntoScanPushdownRule(this);
         }
     }
 
-    public static final CalcIntoScanLogicalRule INSTANCE = new CalcIntoScanLogicalRule(Config.DEFAULT);
+    public static final CalcIntoScanPushdownRule INSTANCE = new CalcIntoScanPushdownRule(Config.DEFAULT);
 
-    private CalcIntoScanLogicalRule(Config config) {
+    private CalcIntoScanPushdownRule(Config config) {
         super(config);
     }
 
     @Override
     public void onMatch(RelOptRuleCall call) {
-        CalcLogicalRel calc = call.rel(0);
-        FullScanLogicalRel scan = call.rel(1);
+        Calc calc = call.rel(0);
+        FullScan scan = call.rel(1);
 
         HazelcastTable table = OptUtils.extractHazelcastTable(scan);
-
         RexProgram program = calc.getProgram();
+        assert scan.getConvention() == LOGICAL; // support it for physical rels also.
 
         List<RexNode> newProjects = program.expandList(program.getProjectList());
         HazelcastTable newTable = table.withProject(newProjects, program.getOutputRowType());
 
         if (program.getCondition() != null) {
-            newTable = newTable.withFilter(program.expandLocalRef(program.getCondition()));
+            RexNode calcFilter = program.expandLocalRef(program.getCondition());
+            RexNode tableFilter = table.getFilter();
+            newTable = newTable.withFilter(
```

Comment:
We need to not only apply the filter, but also merge the projections.

Reply:
I didn't found any way to do it yet, but I merge filters already.
Actually, the PR header highlights it.